### PR TITLE
Adds course completion event and user identification by email

### PR DESCRIPTION
**Pull request by CinecaElearning**
The changes implemented here are about the following issues:
https://github.com/LearningLocker/xAPI-Recipe-Emitter/issues/29
https://github.com/LearningLocker/xAPI-Recipe-Emitter/issues/30
https://github.com/LearningLocker/xAPI-Recipe-Emitter/issues/31

### CODE REVIEW

File: src/Events/CourseCompleted.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    /**
+     * Reads data for an event.
+     * @param [String => Mixed] $opts
+     * @return [String => Mixed]
+     * @override Event
+     */
+    public function read(array $opts) {
+        return array_merge(parent::read($opts), [
+            'object' => $this->readCourse($opts),
```
Comment (ryansmith94) [inline src/Events/CourseCompleted.php]: Extend CourseViewed class to remove this line.

File: src/Events/Event.php [hunk 2]
```diff
@@ -55,13 +55,21 @@ public function read(array $opts) {
...
-                'name' => $opts[$key.'_id'],
-            ],
-        ];
+		if (isset($opts[$key.'_email']) && $opts[$key.'_email']) {	// priority to mbox, then account
+			return [
+				'name' => $opts[$key.'_name'],
+				'mbox' => $opts[$key.'_email'],
+			];
+		}
+		else {
```
Comment (ryansmith94) [inline src/Events/Event.php]: `else` should be on the same line as the `if` closing brace as per the style guide.

File: src/Events/Event.php [hunk 3]
```diff
@@ -55,13 +55,21 @@ public function read(array $opts) {
...
 
     protected function readUser(array $opts, $key) {
-        return [
-            'name' => $opts[$key.'_name'],
-            'account' => [
-                'homePage' => $opts[$key.'_url'],
-                'name' => $opts[$key.'_id'],
-            ],
-        ];
+		if (isset($opts[$key.'_email']) && $opts[$key.'_email']) {	// priority to mbox, then account
```
Comment (ryansmith94) [inline src/Events/Event.php]: Not sure we should "prioritise" the `email` for the sake of data protection. I think it's probably best for the translator to pass through an additional option that explicitly specifies which identifier to use.

File: src/Repository.php [hunk 4]
```diff
@@ -21,7 +21,9 @@ public function __construct(TinCanRemoteLrs $store) {
      * @return [string => mixed]
      */
     public function createEvents(array $statements) {
-        $this->store->saveStatements($statements);
+        $LRSResponse = $this->store->saveStatements($statements);
```
Comment (ryansmith94) [inline src/Repository.php]: As I mentioned in #31, I'm not a fan of this. I think we can do better and this should be discussed in #26.

File: tests/CourseCompletedTest.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+     * @override EventTest
+     */
+    public function setup() {
+        $this->event = new Event($this->repo);
+    }
+
+    protected function constructInput() {
+        return array_merge(
+            parent::constructInput(),
+            $this->contructObject('course')
```
Comment (ryansmith94) [inline tests/CourseCompletedTest.php]: Can probably extend the CourseViewedTest to remove this line.

File: src/Events/Event.php (hunk 2 above)
Comment (CinecaElearning) [inline src/Events/Event.php]: Done

File: tests/CourseCompletedTest.php (hunk 5 above)
Comment (CinecaElearning) [inline tests/CourseCompletedTest.php]: I tried that as you suggested. The test gives a failure when we launch parent::assertOutput which tries to assert the view verb using the completed verb.

File: src/Events/Event.php (hunk 3 above)
Comment (CinecaElearning) [inline src/Events/Event.php]: Done

File: src/Events/CourseCompleted.php (hunk 1 above)
Comment (CinecaElearning) [inline src/Events/CourseCompleted.php]: Done

**Comment by davidpesce** [comment]
The remaining relevant chunk of code (email as identifier) has been included in a separate PR #39. Closing this as it's no longer needed.

**Pull request closed without merging**